Sms #185

kuanb opened the pull request:
Testing additions, plenty of fixes. :(

kuanb commented on app/controllers/sms.js:
```diff
@@ -24,61 +24,62 @@ module.exports = {
+              }
+              if (content) {
+                // Switch to only send when app is in production
+                if (process.env.CCENV !== "testing") {
```
move this to lowest level possible, so in model

kuanb replied:
done

kuanb commented:
@maxmcd this is at a point where you could review and allow to merge or not. i think i fixed some "blob" issues and now that sms logic is good, i could start other efforts and features on new branches with new issues.

kuanb commented:
so just to say this is awaiting an "lgtm"

maxmcd commented on app/models/departments.js:
```diff
@@ -4,12 +4,119 @@
+          active: true
+        })
+      .then(() => {
+        fulfill()
```
Return an instance of the thing you create. 

Actually, we have an instance of `create` on the BaseModel that returns an instance. If you want create methods that take attributes it might be better to do it like this: 

```javascript
static createOne (orgID, name, phoneNumber, userID) {
  return new Promise((fulfill, reject) => {
    return this.create({
      organization: orgID,
      name: name,
      phone_number: phoneNumber,
      created_by: userID,
      active: true
    })
  }
}
```

if there was a naming conflict, which there likely should be as we would want all our create methods to have the same name. So if the method on BaseModel is called createOne we could call `super.createOne` instead of `this.create`

maxmcd commented on app/models/departments.js:
```diff
@@ -4,12 +4,119 @@
+      db("departments")
+        .where("department_id", departmentID)
+        .update({
+          name: name,
```
Same deal with the create. We don't have an edit method in BaseModel, but we should probably create one, and use it to build these methods.

kuanb replied:
Same deal... In terms of returning the department created? Or in terms of building out a basic model method and using that instead. I think the latter is a thing we should do that I will segment. But the first thing is something I can do now for this PR

maxmcd replied:
Both are true. And yeah, I just made an update method in the voice branch for BaseModel, so might be worth it to compare notes and fix that later

maxmcd commented on test/testSMSData.js:
```diff
@@ -0,0 +1,21 @@
+module.exports = {
+    ToCountry: "US",
+    ToState: "CA",
+    SmsMessageSid: "SMd72ee1dd559b20b05a9a005df946b32f",
```
data goes in the test/data folder, might not have merged that yet, but feel free to create it on this branch

kuanb replied:
yeah was not there but is now

maxmcd commented on test/app/smsController.js:
```diff
@@ -0,0 +1,120 @@
+
+const twilioAgent = supertest.agent(APP);
+const smsData = require('../testSmsData');
+
```
Seems like good controller test coverage here. You have a bunch of well-defined model methods now that are used to build these controllers. Would ideally want to see model tests for those methods as well. These would arguably be easier to build/validate and could have been build in tandem with the refactor to test that methods were working without hitting the controller with a request. Tighter feedback loops, more granular test cases, etc....

kuanb replied:
So the takeaway here is add model level tests to pair with these controller level tests. Will do.

kuanb commented on app/models/capture.js:
```diff
@@ -61,41 +64,58 @@ class CaptureBoard {
+      }).then((resp) => {
+
+        // Query 2: Close all other conversation that client has open
+        return db("convos")
```
roger, will do. `associateConversation` is a sort of mess and im just leaving it for another day to clean up

kuanb commented on app/models/commConns.js:
```diff
@@ -19,6 +33,35 @@ class CommConns {
     })
   }
+
+  static findByCommId (communicationId) {
```
added `findManyByAttribute` but yes as we discussed need to also be able to add the whole  logic that supports `.and.where("retired", null)`

kuanb replied:
Went ahead and updated such that:

```
  static findManyByAttribute(attributeName, value, otherOperations) {
    if (!otherOperations) {
      otherOperations = (dbCall) => {
        return dbCall;
      }
    }
    this._checkModelValidity();
    return new Promise((fulfill, reject) => {
      let basicDbCall = db(this.tableName)
                          .where(attributeName, value);
      otherOperations(basicDbCall)
      .then((objects) => {
        return this._getMultiResponse(objects, fulfill)
      }).catch(reject)
    })
  }
```

Is in models and it looks like this when calling from CommConns:

```
      CommConns.findManyByAttribute("comm", 1, (dbCall) => {
        return dbCall
          .andWhere("retired", null);
      })
      .then((commConns) => {
        clients.length.should.be.exactly(commConns.length);
        done();
      }).catch(done);```

maxmcd replied:
this is great

Would make sure to try and not actually use `findManyByAttribute` in real code. Better to define a wrapper method that calls in on the model with a better name. 

So in that test example would prefer that you're calling `CommConns.findManyByCommId` which in turn calls the method you've written above

kuanb commented on app/models/conversations.js:
```diff
@@ -25,18 +23,186 @@ class Conversations extends BaseModel {
+            open: false
+          });
+      }).then(() => {
+        fulfill();
```
Modified 


```        return db("convos")
          .whereIn("convid", conversationIds)
          .update({
            accepted: false,
            open: false
          })
          .returning("*");
      }).then((conversations) => {
        this._getMultiResponse(conversations, fulfill);
      }).catch(reject);```

kuanb commented:
```
  65 passing (2s)
  5 pending```

Hope travis works out. Added testing for all the models in sms in lib. Also added the functionality of FindByAttr to have the ability to add more where clauses or other manipulation... One issue remains, you can't access the `db` object which makes doing something like a leftjoin hard, but perhaps that is ok because we want to make complex calls like that their own verbose method and not just something janky we are slapping on to the find by attr func?

otherwise, lemme know if it looks lgtm

maxmcd commented:
LGTM
